docs/getting-started/walkthroughs/datastream_api.md

```diff
@@ -0,0 +1,897 @@
+public class FraudDetector extends KeyedProcessFunction<Long, Transaction, Alert> {
+
+    public static final double SMALL_AMOUNT = 0.01;
+
+    public static final double LARGE_AMOUNT = 500.00;
+
+    public static final long ONE_DAY = 24 * 60 * 60 * 1000;
+
+    @Override
+    public void processElement(
+        Transaction transaction,
+        Context context,
+        Collector<Alert> collector) throws Exception {
+  
+        Alert alert = new Alert();
+        alert.setId(transaction.getAccountId());
+
+        collector.collect(alert);
+    }
+}
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+object FraudDetector {
+
+    val SMALL_AMOUNT = 0.01
+
+    val LARGE_AMOUNT = 500.00
+
+    val ONE_DAY = 24 * 60 * 60 * 1000L
+}
+
+class FraudDetector extends KeyedProcessFunction[Long, Transaction, Alert] {
+
+    override def processElement(
+        transaction: Transaction,
+        context: Context,
+        collector: Collector[Alert]): Unit = {
+
+        Alert alert = new Alert
+        alert.setId(transaction.getAccountId)
+
+        collector.collect(alert)
+    }
+}
+{% endhighlight %}
+</div>
+</div>
+
+## Writing An Initial Application 
+
+For the initial implementation, the fraud detector should output an alert for any account that makes a small transaction immediately followed by a large one. Where small is anything less than $0.10 and large is more than $500.
+Imagine your fraud detector processes the following stream of transactions for a particular account.
+
+<p class="text-center">
+    <img alt="Transactions" width="80%" src="{{ site.baseurl }}/fig/fraud-transactions.svg"/>
+</p>
+
+Transactions 3 and 4 should be marked as fraudulent because it is a small transaction, $0.09, followed by a large one, $510.
+Alternatively, transactions 7, 8, and 9 are not fraud because the small amount of $0.02 is not immediately followed by the large one; instead, there is an intermediate transaction that breaks the pattern.
+
+To do this, the fraud detector must _remember_ information across events; a large transaction is only fraudulent if the previous one was small.
+Remembering information across events requires [state]({{ site.baseurl }}/concepts/glossary.html#managed-state) and so you will implement your fraud detector using a [KeyedProcessFunction]({{ site.baseurl }}/dev/stream/operators/process_function.html), which provides fine-grained control over state and time.
+
+The most straightforward implementation would be to set a flag whenever a small transaction is processed.
+This way, when a large transaction comes through, you can check if the flag is set for that account.
+If it is, then this is fraud and output an alert.
+This flag is what you want to store in Flink state.
+
+The most basic type of state in Flink is [ValueState]({{ site.baseurl }}/dev/stream/state/state.html#using-managed-keyed-state), a data type that provides _fault tolerant_, _managed_, _per key_ state.
```

Comment:
Do you think, we need more emphasis on the **per key**  part since most people don't get this at first? I see, you're getting to this below, but maybe still...

Reply:
Agreed, I decided to remove the reference to `_per key_` from the intro and just focus on the fault tolerance and then introduce and expand on per-key below.